File: 0141-linked-list-cycle/0141-linked-list-cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        tortoise=head
        rabbit=head
        while rabbit and rabbit.next:
            tortoise=tortoise.next
            rabbit=rabbit.next.next
            if tortoise==rabbit:
                return True
        return False
        

File: 0141-linked-list-cycle/0141-linked-list-cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        ListNode tortoise =head;
        ListNode rabbit=head;
        while(rabbit!=null && rabbit.next!=null){
            tortoise=tortoise.next;
            rabbit=rabbit.next.next;
            if(tortoise==rabbit) return true;
        }
        return false;
    }
}